tests/acceptance/30_custom_promise_types/example_module.sh

```diff
@@ -0,0 +1,148 @@
+reset_state() {
+    # Set global variables before we begin another request
+
+    # Variables parsed directly from request:
+    request_operation=""
+    request_log_level=""
+    request_promise_type=""
+    request_promiser=""
+    request_attribute_message=""
+
+    # Variables to put into response:
+    response_result=""
+
+    # Other state:
+    saw_unknown_key="no"
+    saw_unknown_attribute="no"
+}
+
+handle_input_line() {
+
+    # Split the line of input on the first '=' into 2 - key and value
+    IFS='=' read -r key value <<< "$1"
+    
+    case "$key" in 
+    operation)
+        request_operation="$value" ;;
+    log_level)
+        request_log_level="$value" ;;
+    promise_type)
+        request_promise_type="$value" ;;
+    promiser)
+        request_promiser="$value" ;;
+    attribute_message)
+        request_attribute_message="$value" ;;
+    attribute_*)
+        attribute_name=${key#"attribute_"}
+        echo "log_error=Unknown attribute: '$attribute_name'"
+        saw_unknown_attribute="yes" ;;
+    *)
+        saw_unknown_key="yes" ;;
+    esac
+}
+
+receive_request() {
+    # Read lines from input until empty line
+    # Call handle_input_line for each non-empty line
+    while IFS='$\n' read -r line; do
+        if [ "x$line" = "x" ] ; then
+            break
+        fi
+        handle_input_line "$line" # Parses a key=value pair
+    done
+}
+
+write_response() {
+    echo "operation=$request_operation"
+    echo "result=$response_result"
+    echo ""
+}
+
+operation_terminate() {
+    response_result="success"
+    write_response
+    exit 0
+}
+
+operation_validate() {
+    response_result="valid"
+    if [ "$saw_unknown_attribute" != "no" ] ; then
+        response_result="invalid"
+    fi
+
+    if [ "$request_promiser" = "" ] ; then
+        echo "log_error=Promiser must be non-empty"
+        response_result="invalid"
+    fi
+
+    if [ "$request_attribute_message" = "" ] ; then
```

Comment:
`-z "$request_attribute_message"` would be more "bash-way"

Reply:
I guess my way is more portable and more readable?

Reply:
Not sure about portability, but could be more readable for non-bash-believers. :+1:

Reply:
`man test` for the win eh?